internal/command/cmd.go

```diff
@@ -36,75 +37,97 @@ const (
-		cmd.Dir = c.Dir
-		cmd.Stdout = c.Stdout
-		cmd.Stderr = c.Stderr
-		err = cmd.Run()
-		if err == nil {
+		backoff := getNewBackoff(i)
+		log.Info().Msgf("Next retry in %ds...", backoff)
+		time.Sleep(time.Duration(backoff) * time.Second)
+
+		if err = c.execute(i, numOfRetries, commandTimeout); err == nil {
 			log.Info().Msgf("The %s was successful on %d retry", c.Command, i)
 			return nil
 		}
 		log.Warn().Msgf("Error encountered while executing %s : %v", c.Command, err)
-		backoff := 10 * i
-		log.Info().Msgf("Next retry in %ds...", backoff)
-		time.Sleep(time.Duration(backoff) * time.Second)
 	}
 	log.Error().Msgf("Command %s was not successful after %d retries", c.Command, numOfRetries)
 	return err
 }
 
-// RetryCommandWithOutput will retry the given command, every 5 sec until either successful or reached numOfRetries
-// returns (nil, error) if all retries fail, (output, nil) otherwise
-func (c Cmd) RetryCommandWithOutput(numOfRetries int) ([]byte, error) {
-	// have an initial backoff before trying again.
-	log.Info().Msgf("Next retry in %ds...", 15)
-	time.Sleep(15 * time.Second)
-
+// RetryCommandWithOutput retries the given command, with exponential backoff, maxing at 5 min, for numOfRetries times.
+// commandTimeout is used to terminate the command after specified time, regardless if it was successful or not.
+// This prevents commands to be executing indefinitely.
+// returns (nil, error) if all retries fail, (output, nil) otherwise.
+func (c *Cmd) RetryCommandWithOutput(numOfRetries, commandTimeout int) ([]byte, error) {
 	var err error
+	var out []byte
 	for i := 1; i <= numOfRetries; i++ {
-		log.Warn().Msgf("Retrying command %s... (%d/%d)", c.Command, i, numOfRetries)
-		cmd := exec.Command("bash", "-c", c.Command)
-		cmd.Dir = c.Dir
-		cmd.Stdout = c.Stdout
-		cmd.Stderr = c.Stderr
-
-		var out []byte
-		out, err = cmd.CombinedOutput()
-		if err == nil {
+		backoff := getNewBackoff(i)
+		log.Info().Msgf("Next retry in %ds...", backoff)
+		time.Sleep(time.Duration(backoff) * time.Second)
+
+		if out, err = c.executeWithOutput(i, numOfRetries, commandTimeout); err == nil {
 			log.Info().Msgf("The %s was successful after %d retry", c.Command, i)
 			return out, nil
 		}
-
 		log.Warn().Msgf("Error encountered while executing %s : %v", c.Command, err)
-		backoff := 10 * i
-		log.Info().Msgf("Next retry in %ds...", backoff)
-		time.Sleep(time.Duration(backoff) * time.Second)
 	}
 	log.Error().Msgf("Command %s was not successful after %d retries", c.Command, numOfRetries)
 	return nil, err
 }
 
-// GetStdOut returns an io.Writer for exec with the defined prefix
-// Cannot be used with cmd.CombinedOutput()
+// execute executes the cmd with context canceled after commandTimeout seconds.
+// Returns error if unsuccessful, nil otherwise.
+func (c *Cmd) execute(i, numOfRetries, commandTimeout int) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(commandTimeout)*time.Second)
```

Comment:
I'm not sure how do I feel about the timeout, the time that the command will take to execute depend on the size of the cluster and we can't really make any assumptions about that.

I've seen that you've set the timeout to 600 seconds (10 mins) if I request to create a large enough cluster there is a certainty that it won't be build due to exceeding the timeout threashold

Reply:
That's a good point. I think we have two options here

1. Disregard the `commandTimeout` for calls which are affected by the size/complexity of a cluster
2. Somehow scale the `commandTimeout` based on the size/complexity of a cluster

I lean toward the second option since we can scale that number by node pool size/number of node pools, however, I want to hear other opinions too.

Reply:
if we wanted to keep the timeout, then we could benchmark clusters with different sizes to get the idea of how long does it take to execute each command and maybe base the number off of that, but I'm not sure how accurate would this be

Reply:
This look much better, I think its fine for now 👍 , I agree with the timeout on kubectl commands which we can increase in the future if it proves to be a problem